### PR TITLE
Feature/event listener

### PR DESCRIPTION
**Pull request by rashidsp**

### CODE REVIEW

File: lib/optimizely/notification_center.rb [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+#
+module Optimizely
+  class NotificationCenter
+    attr_reader :notifications
+    attr_reader :notification_id
+
+    NOTIFICATION_TYPES = {
+      ACTIVATE: 'ACTIVATE: experiment, user_id, attributes, variation, event',
+      TRACK: 'TRACK: event_key, user_id, attributes, event_tags, event',
+      FEATURE_ROLLOUT: 'FEATURE_ROLLOUT:feature_key, user_id, attributes, audiences',
```
Comment (mikeng13) [inline lib/optimizely/notification_center.rb]: As mentioned in our meeting today, please remove `FEATURE_ROLLOUT` and `FEATURE_EXPERIMENT` notification types for now.

Comment (rashidsp) [reply lib/optimizely/notification_center.rb]: Removed.

File: lib/optimizely.rb [hunk 2]
```diff
@@ -231,6 +234,10 @@ def track(event_key, user_id, attributes = nil, event_tags = nil)
       rescue => e
         @logger.log(Logger::ERROR, "Unable to dispatch conversion event. Error: #{e}")
       end
+      @notification_center.fire_notifications(
```
Comment (thomaszurkan-optimizely) [inline lib/optimizely.rb]: please change to send_notifications

File: lib/optimizely/notification_center.rb [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+      # Adds notification callback to the notification center
+
+      # Args:
+      #  notification_type: one of the constants in NOTIFICATION_TYPES
+      #  notification_callback: function to call when the event is fired
+
+      # Returns:
+      #  notification ID used to remove the notification
+
+      return nil unless notification_type_valid?(notification_type)
```
Comment (thomaszurkan-optimizely) [inline lib/optimizely/notification_center.rb]: I actually allow any notification to be added.  So, a user could use the Optimizely.NotificationCenter to hold any notification type.

File: lib/optimizely/notification_center.rb [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+      @notifications[notification_type] = []
+      @logger.log Logger::INFO, "All callbacks for notification type #{notification_type} have been removed."
+    end
+
+    def clean_all_notifications
+      # Removes all notifications
+      @notifications.keys.each { |key| @notifications[key] = [] }
+    end
+
+    def fire_notifications(notification_type, *args)
```
Comment (thomaszurkan-optimizely) [inline lib/optimizely/notification_center.rb]: change to send_notifications.

File: lib/optimizely/notification_center.rb (hunk 3 above)
Comment (rashidsp) [reply lib/optimizely/notification_center.rb]: I don't understand, we are only sending notifications from NOTIFICATION_TYPES Constant!
Then what is the purpose of having non-functional notification types.

File: lib/optimizely.rb (hunk 2 above)
Comment (rashidsp) [reply lib/optimizely.rb]: Updated

File: lib/optimizely/notification_center.rb (hunk 4 above)
Comment (rashidsp) [reply lib/optimizely/notification_center.rb]: Updated

**Comment by mikeng13** [comment]
E2E tests are passing @thomaszurkan-optimizely lemme know if you want to take a look at this before I merge it in!

**Pull request merged**